### PR TITLE
feat: define and implement a structure to output results.

### PR DESCRIPTION
**Pull request by kobayu858**
- Define and implement a structure to output results.
- Output functions not yet defined.
- Set value in structure in schedule.

### CODE REVIEW

File: lib/src/fixed_priority_scheduler.rs [hunk 1]
```diff
@@ -4,18 +4,22 @@ use crate::{
...
 const DUMMY_EXECUTION_TIME: i32 = 1;
 
+#[derive(Clone, Default)]
 pub struct FixedPriorityScheduler<T>
 where
     T: ProcessorBase + Clone,
 {
     pub dag: Graph<NodeData, i32>,
     pub processor: T,
+    pub node_log: Vec<NodeLog>,
```
Comment (atsushi421) [inline lib/src/fixed_priority_scheduler.rs:21]:
```suggestion
    pub node_logs: Vec<NodeLog>,
```

File: lib/src/fixed_priority_scheduler.rs [hunk 2]
```diff
@@ -26,15 +30,27 @@ where
         Self {
             dag: dag.clone(),
             processor: processor.clone(),
+            node_log: vec![NodeLog::default(); dag.node_count()],
+            processor_log: ProcessorLog {
+                core_log: vec![CoreLog::default(); processor.get_number_of_cores()],
```
Comment (atsushi421) [inline lib/src/fixed_priority_scheduler.rs:35]:
```suggestion
                core_logs: vec![CoreLog::default(); processor.get_number_of_cores()],
```

File: lib/src/scheduler.rs [hunk 3]
```diff
@@ -11,3 +11,51 @@ where
     fn set_processor(&mut self, processor: &T);
     fn schedule(&mut self) -> (i32, VecDeque<NodeIndex>);
 }
+
+#[derive(Clone, Default)]
+pub struct NodeLog {
+    pub core_id: usize,
+    pub node_id: i32,
+    pub start_time: i32,
+    pub end_time: i32,
```
Comment (atsushi421) [inline lib/src/scheduler.rs:20]: In real-time scheduling domain, `finish_time` is more common.

File: lib/src/scheduler.rs [hunk 4]
```diff
@@ -11,3 +11,51 @@ where
...
+}
+
+#[derive(Clone, Default)]
+pub struct ProcessorLog {
+    pub core_log: Vec<CoreLog>,
```
Comment (atsushi421) [inline lib/src/scheduler.rs:25]:
```suggestion
    pub core_logs: Vec<CoreLog>,
```

File: lib/src/scheduler.rs [hunk 5]
```diff
@@ -11,3 +11,51 @@ where
...
+    pub average_utilization_rate: f32,
+    pub variance_utilization_rate: f32,
+}
+
+impl ProcessorLog {
+    pub fn set_average_utilization_rate(&mut self) {
+        self.average_utilization_rate = self
+            .core_log
+            .iter()
+            .map(|core_data| core_data.utilization_rate)
```
Comment (atsushi421) [inline lib/src/scheduler.rs:35]:
```suggestion
            .map(|core_log| core_log.utilization_rate)
```

File: lib/src/scheduler.rs [hunk 6]
```diff
@@ -11,3 +11,51 @@ where
...
+            .sum::<f32>()
+            / self.core_log.len() as f32;
+    }
+
+    pub fn set_variance_utilization_rate(&mut self) {
+        self.variance_utilization_rate = self
+            .core_log
+            .iter()
+            .map(|core_data| (core_data.utilization_rate - self.average_utilization_rate).powi(2))
```
Comment (atsushi421) [inline lib/src/scheduler.rs:44]:
```suggestion
            .map(|core_log| (core_log.utilization_rate - self.average_utilization_rate).powi(2))
```

File: lib/src/scheduler.rs [hunk 7]
```diff
@@ -11,3 +11,51 @@ where
...
+
+#[derive(Clone, Default)]
+pub struct CoreLog {
+    pub core_id: usize,
+    pub total_proc_time: i32,
+    pub utilization_rate: f32,
+}
+
+impl CoreLog {
+    pub fn set_utilization_rate(&mut self, schedule_length: i32) {
```
Comment (atsushi421) [inline lib/src/scheduler.rs:58]: A setter is simply a method that assigns a value to a field.
https://qiita.com/takahirocook/items/27828bc8477735612021

This method is not a setter, so a name like `calculate_utilization` is preferred.

File: lib/src/scheduler.rs [hunk 8]
```diff
@@ -11,3 +11,51 @@ where
...
+
+#[derive(Clone, Default)]
+pub struct ProcessorLog {
+    pub core_log: Vec<CoreLog>,
+    pub average_utilization_rate: f32,
+    pub variance_utilization_rate: f32,
+}
+
+impl ProcessorLog {
+    pub fn set_average_utilization_rate(&mut self) {
```
Comment (atsushi421) [inline lib/src/scheduler.rs:31]:
```suggestion
    pub fn calculate_average_utilization(&mut self) {
```

File: lib/src/fixed_priority_scheduler.rs [hunk 9]
```diff
@@ -26,15 +30,27 @@ where
         Self {
             dag: dag.clone(),
             processor: processor.clone(),
+            node_log: vec![NodeLog::default(); dag.node_count()],
+            processor_log: ProcessorLog {
```
Comment (atsushi421) [inline lib/src/fixed_priority_scheduler.rs:34]: It is more natural to implement the new method and assign the core_id here.

**Review by atsushi421 — CHANGES_REQUESTED** [review]
The `utilization_rate` should simply be `utilization`.

The variable name `task` is less favourable, because it is difficult to know whether it is an id, node_index or NodeData.

File: lib/src/scheduler.rs (hunk 6 above)
Comment (kobayu858) [reply lib/src/scheduler.rs:44]: Modified.

File: lib/src/scheduler.rs (hunk 5 above)
Comment (kobayu858) [reply lib/src/scheduler.rs:35]: Modified.

File: lib/src/scheduler.rs (hunk 4 above)
Comment (kobayu858) [reply lib/src/scheduler.rs:25]: Modified.

File: lib/src/scheduler.rs (hunk 3 above)
Comment (kobayu858) [reply lib/src/scheduler.rs:20]: Modified.

File: lib/src/fixed_priority_scheduler.rs (hunk 2 above)
Comment (kobayu858) [reply lib/src/fixed_priority_scheduler.rs:35]: Modified.

File: lib/src/fixed_priority_scheduler.rs (hunk 1 above)
Comment (kobayu858) [reply lib/src/fixed_priority_scheduler.rs:21]: Modified.

File: lib/src/scheduler.rs (hunk 7 above)
Comment (kobayu858) [reply lib/src/scheduler.rs:58]: Modified.

File: lib/src/scheduler.rs (hunk 8 above)
Comment (kobayu858) [reply lib/src/scheduler.rs:31]: Modified.

File: lib/src/fixed_priority_scheduler.rs (hunk 1 above)
Comment (kobayu858) [reply lib/src/fixed_priority_scheduler.rs:21]: Modified.

File: lib/src/fixed_priority_scheduler.rs (hunk 2 above)
Comment (kobayu858) [reply lib/src/fixed_priority_scheduler.rs:35]: Modified.

File: lib/src/fixed_priority_scheduler.rs (hunk 9 above)
Comment (kobayu858) [reply lib/src/fixed_priority_scheduler.rs]: Add `new()` function to each structure
In addition, the id of node and core are changed to be set to a value at the time of `new()`.

**Review by atsushi421 — APPROVED** [review]
LGTM

**Pull request merged**